src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -240,15 +250,27 @@ internal JsonConverter GetConverterInternal(Type typeToConvert)
                 return _cachingContext.GetOrAddConverter(typeToConvert);
             }
 
-            return GetConverterFromType(typeToConvert);
+            Debug.Assert(s_converterRetrievalFunc != null);
+            return s_converterRetrievalFunc(typeToConvert, this);
         }
 
-        private JsonConverter GetConverterFromType(Type typeToConvert)
+        private JsonConverter? GetConverterForTypeFromContext(Type typeToConvert)
+            => _serializerContext?.GetTypeInfo(typeToConvert)?.PropertyInfoForTypeInfo?.ConverterBase;
+
+        private JsonConverter GetConverterForTypeSourceGen(Type typeToConvert)
+        {
+            JsonConverter? converter = GetConverterForTypeFromContext(typeToConvert);
+            Debug.Assert(converter != null);
```

Comment:
I don't see what guarantees this... should this be using reflection path here as fallback?